core/base/perturbation.cpp

```diff
@@ -0,0 +1,107 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/perturbation.hpp>
+
+
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+
+
+template <typename ValueType>
+void Perturbation<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
+{
+    // x = (I + scalar * basis * projector) * b
+    // temp = projector * b                 : projector->apply(b, temp)
+    // x = b                                : x->copy_from(b)
+    // x = 1 * x + scalar * basis * temp    : basis->apply(scalar, temp, 1, x)
+    using vec = gko::matrix::Dense<ValueType>;
+    auto exec = this->get_executor();
+    if (cache_.one == nullptr) {
+        cache_.one = initialize<vec>({gko::one<ValueType>()}, exec);
+    }
+    auto intermediate_size =
+        gko::dim<2>(projector_->get_size()[0], b->get_size()[1]);
+    if (cache_.intermediate == nullptr ||
+        cache_.intermediate->get_size() != intermediate_size) {
+        cache_.intermediate = vec::create(exec, intermediate_size);
+    }
+    projector_->apply(b, lend(cache_.intermediate));
+    x->copy_from(b);
+    basis_->apply(lend(scalar_), lend(cache_.intermediate), lend(cache_.one),
+                  x);
+}
+
+
+template <typename ValueType>
+void Perturbation<ValueType>::apply_impl(const LinOp *alpha, const LinOp *b,
+                                         const LinOp *beta, LinOp *x) const
+{
+    // x = alpha * (I + scalar * basis * projector) b + beta * x
+    //   = beta * x + alpha * b + alpha * scalar * basis * projector * b
+    // temp = projector * b     : projector->apply(b, temp)
+    // x = beta * x + alpha * b : x->scale(beta),
+    //                            x->add_scaled(alpha, b)
+    // x = x + alpha * scalar * basis * temp
+    //                          : basis->apply(alpha * scalar, temp, 1, x)
+    using vec = gko::matrix::Dense<ValueType>;
+    auto exec = this->get_executor();
+    if (cache_.one == nullptr) {
+        cache_.one = initialize<vec>({gko::one<ValueType>()}, exec);
+    }
+    auto intermediate_size =
+        gko::dim<2>(projector_->get_size()[0], b->get_size()[1]);
+    if (cache_.intermediate == nullptr ||
+        cache_.intermediate->get_size() != intermediate_size) {
+        cache_.intermediate = vec::create(exec, intermediate_size);
+    }
+    projector_->apply(b, lend(cache_.intermediate));
+    auto vec_x = as<vec>(x);
+    vec_x->scale(beta);
+    vec_x->add_scaled(alpha, b);
+    if (cache_.alpha_scalar == nullptr) {
+        cache_.alpha_scalar = vec::create(exec, gko::dim<2>(1));
+    }
+    cache_.alpha_scalar->copy_from(alpha);
+    cache_.alpha_scalar->scale(lend(scalar_));
```

Comment:
why are you copying `alpha` in an intermediate instead of using it directly in the `apply` (especially since you are using it in [line 92](https://github.com/ginkgo-project/ginkgo/pull/334/files#diff-b949b55dcf76da1f135f8f545404df50R92) directly)?

If you are worried about it being on the wrong executor, the `make_temporary_clone` in the implementation of [apply in lin_op.hpp](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/lin_op.hpp#L188) takes care of that.

Reply:
My bad, did not read carefully enough. This part is perfectly fine.

Reply:
I use apply to handle this operation. `alpha->apply(scalar_, cache.alpha_scalar)`